### DISCUSSION TITLE
TypeError: Cannot read property 'execPost' of undefined

### DISCUSSION THREAD

**Discussion by mrdulin** [Automattic/mongoose#13570] [Q&A]
Try to use the `Query` class to build a query. 

```ts
import mongoose from 'mongoose';
import { config } from '../../config';

mongoose.set('debug', true);

const schema = new mongoose.Schema({
	name: String,
});
const User = mongoose.model('user', schema);

(async function main() {
	try {
		await mongoose.connect(config.MONGODB_URI);

		// test
		const query = new mongoose.Query();
		const r2 = await query
			.setOptions({ lean: true })
			// @ts-ignore
			.collection(User.collection)
			.where({ name: 'Tom Benzamin' })
			.select({ _id: 0, address: 0 })
			.exec();
		console.log('r2: ', r2);
	} catch (error) {
		console.error(error);
	} finally {
		await mongoose.connection.close();
	}
})();
```

Got a runtime error:

```zsh
TypeError: Cannot read property 'execPost' of undefined
    at /home/lindu/workspace/mongoose-codelab/v7/node_modules/mongoose/lib/query.js:4440:28
    at new Promise (<anonymous>)
    at _executePostHooks (/home/lindu/workspace/mongoose-codelab/v7/node_modules/mongoose/lib/query.js:4437:10)
    at Query.exec (/home/lindu/workspace/mongoose-codelab/v7/node_modules/mongoose/lib/query.js:4409:15)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async main (/home/lindu/workspace/mongoose-codelab/v7/stackoverflow/53437031/index.ts:20:14)
```

The code line that throws this error is [here](https://github.com/Automattic/mongoose/blob/7.3.1/lib/query.js#L4440)

```ts
function _executePostHooks(query, res, error, op) {
  return new Promise((resolve, reject) => {
    const opts = error ? { error } : {};

    query._queryMiddleware.execPost(op || query.op, query, [res], opts, (error, res) => {
      if (error) {
        return reject(error);
      }

      resolve(res);
    });
  });
}
```

Why `query._queryMiddleware` is `undefined`? Did I build the query correctly? There is no post middleware in my code.

**Comment by vkarpov15** [comment]
We will improve this error message, but the general idea is that you're not supposed to instantiate the `Query` class directly. From the Query API docs:

![image](https://github.com/Automattic/mongoose/assets/1620265/6fdd0fc2-43ce-41cd-b93c-3b28407b971b)

Why are you using this pattern to create queries rather than using models?

**Reply by mrdulin** [reply]
Yeah, I know it's recommended to use the static methods of the Model. Shouldn't these two ways be equivalent? `Model.find()` is equivalent to `new mongoose.Query()`? If instantiating the `Query` class directly is not recommended, why not make it private? Anyway, is this error caused by a bug?

**Reply by hasezoey** [reply]
> Anyway, is this error caused by a bug?

no, the query is just not expected to execute without a model and operation being associated with it (the error does not directly happen at `new` because those can be set before being executed)

you should be able to do
```ts
new Query(undefined, undefined, MODEL, undefined);
// or
const query = new Query();
query.model = MODEL;
```

see 
https://github.com/Automattic/mongoose/blob/67b42244d7da2ec8597e21ea5699915dc3a12a06/lib/model.js#L2077-L2081

also see #13571

**Reply by mrdulin** [reply]
I thought `query.collection(User.collection)` make an association with a model/collection.

**Reply by vkarpov15** [reply]
You can think of Query as an abstract class - the Query class methods are part of our documented API, but the constructor isn't, because we don't expect people to instantiate the Query class directly.
